app/controllers/PaypalController.scala

```diff
@@ -0,0 +1,105 @@
+package controllers
+
+import actions.CommonActions._
+import com.gu.i18n.CountryGroup
+import play.api.data.{Form, FormError}
+import play.api.data.Forms._
+import play.api.libs.ws.WSClient
+import play.api.mvc.{Controller, Result}
+import services.PaymentServices
+import play.api.Logger
+import play.api.data.format.Formatter
+import play.api.libs.json._
+import utils.TransactionUtil
+import play.api.libs.functional.syntax._
+
+import scala.util.Right
+
+
+class PaypalController(ws: WSClient, paymentServices: PaymentServices, transactionUtil :TransactionUtil) extends Controller {
+
+  implicit val countryGroupFormatter = new Formatter[CountryGroup] {
+    type Result = Either[Seq[FormError], CountryGroup]
+
+    override def bind(key: String, data: Map[String, String]): Result = {
+      data.get(key).flatMap(CountryGroup.byId(_)).fold[Result](Left(Seq.empty))(countryGroup => Right(countryGroup))
+    }
+
+    override def unbind(key: String, value: CountryGroup): Map[String, String] = Map(key -> value.id)
+  }
+
+  def executePayment(countryGroup: CountryGroup, paymentId: String, token: String, payerId: String) = NoCacheAction { implicit request =>
+    val paypalService = paymentServices.paypalServiceFor(request)
+    paypalService.executePayment(paymentId, token, payerId) match {
+      case Right(_) => Redirect(routes.Giraffe.thanks(countryGroup).url, SEE_OTHER)
+      case Left(error) => handleError(countryGroup, s"Error executing PayPal payment: $error")
+    }
+  }
+
+  case class PaymentData(
+    countryGroup: CountryGroup,
+    amount: BigDecimal
+  )
+
+  val paypalForm = Form(
+    mapping(
+      "countryGroup" -> of[CountryGroup],
+      "amount" -> bigDecimal(10, 2)
+    )(PaymentData.apply)(PaymentData.unapply)
+  )
+
+  def authorize = NoCacheAction { implicit request =>
+    paypalForm.bindFromRequest().fold[Result](
+      hasErrors = form => handleError(CountryGroup.UK, form.errors.mkString(",")),
+      success = form => {
+        val transactionId = transactionUtil.newTransactionId
```

Comment:
Do you mind if we call it contributionId? that's how I called it in the database.